File: antia_numerical_methods@Beliavsky/c/fun/simpl1.c
/*	To solve a linear programming problem in the standard form arising
		 in L_1 minimisation problems using the simplex method

	A : (input/output) Array of length IA*(N-M+1) containing
		the tableau of simplex algorithm
		A[I+1][0]=c_i, the cost coefficients
		Columns 1 to M contain constraints with A[0][j]=b_j and A[i+1][j]=a_i
	IA : (input) Second dimension of array A as declared in the calling
		function. IA >= M+2
	N : (input) Number of variables, each is constraint to be >=0
	M : (input) Number of constraints of form a^T X = b_i >= 0
	ID : (input/output) Integer array of length M+1 which contains
		information about interchange of variables on LHS
	IV : (input/output) Integer array of length N-M+1 which contains
		information about interchange of variables on RHS
	AEPS : (input) Required accuracy, any coefficient <AEPS, may be
		assumed to be zero
		
	Error status is returned by the value of the function SIMPL1.
		0 value implies successful execution
		63 implies that the objective function is unbounded from below
		635 implies that the simplex algorithm failed to find
			the optimal feasible vector

	Required functions : None
*/

#include <math.h>

int simpl1(double *a, int ia, int n, int m, int id[], int iv[], double aeps)

{
	int i,j,k,k1,l,l1,jf,m1,n1,it, nit=20;
	double rmin,r1,aj;

	jf=0;
	m1=m+1; n1=n-m+1;

/*	The simplex iteration */
	for(it=1; it<=nit*(n+m); ++it) {
/*	Finding the minimum reduced cost coefficient */
		rmin=0.0; k=0;
		for(j=1; j<n1; ++j) {
			if(a[jf+j*ia]<rmin) {rmin=a[jf+j*ia]; k=j;}
			else if(iv[j]<=m-1 && 2-a[jf+j*ia]< rmin) {rmin=2-a[jf+j*ia]; k=-j;}
		}
		if(rmin>= -aeps) return 0;

/*	Finding the pivot element */
		k1=abs(k);
		rmin=0.0;
		l=0;
		for(j=1; j<=m; ++j) {
			aj=a[j+k1*ia];
			if(k<0) aj=-aj;
			if(aj > aeps) {
				r1=a[j]/aj;
				if(r1<rmin || l==0) {rmin=r1; l=j;}
			}
		}
		if(l==0) return 63;  /*	The objective function is unbounded from below */

		if(k<0) {
			a[jf+k1*ia]=-2+a[jf+k1*ia];
			iv[k1]=-iv[k1];
		}

/*	Exchange the variables */
		l1=id[l]; id[l]=iv[k1]; iv[k1]=l1;
		for(j=0; j<n1; ++j) {
			if(j != k1) {
				r1=a[l+j*ia]/a[l+k1*ia];
				for(i=0; i<m1; ++i) {
					if(i != l) a[i+j*ia]=a[i+j*ia]-a[i+k1*ia]*r1;
				}
			}
		}

		r1=fabs(a[l+k1*ia]);
		for(j=0; j<n1; ++j) {
			if(j != k1) a[l+j*ia]=a[l+j*ia]/r1;
		}
		for(i=0; i<m1; ++i) {
			if(i != l) a[i+k1*ia]= -a[i+k1*ia]/a[l+k1*ia];
		}
		a[l+k1*ia]=1./r1;
	}

	return 635;
}

File: antia_numerical_methods@Beliavsky/f77/sub4/simpl1.f
!	To solve a linear programming problem in the standard form arising
!		 in L_1 minimisation problems using the simplex method
!
!	A : (input/output) Real array of length IA*(N-M+1) containing
!		the tableau of simplex algorithm
!		A(1,I+1)=c_i, the cost coefficients
!		Rows 2 to M+1 contain constraints with A(j,1)=b_j and A(j,i+1)=a_i
!	IA : (input) First dimension of array A as declared in the calling
!		program. IA .GE. M+2
!	N : (input) Number of variables, each is constraint to be .GE.0
!	M : (input) Number of constraints of form a^T X = b_i .GE. 0
!	ID : (input/output) integer array of length M+1 which contains
!		information about interchange of variables on LHS
!	IV : (input/output) integer array of length N-M+1 which contains
!		information about interchange of variables on RHS
!	IER : (output) Error parameter, IER=0 implies successful execution
!		IER=63 implies that the objective function is unbounded from below
!		IER=635 implies that the simplex algorithm failed to find
!			the optimal feasible vector
!	AEPS : (input) Required accuracy, any coefficient <AEPS, may be
!		assumed to be zero
!
!	Required routines : None
!
      SUBROUTINE SIMPL1(A,IA,N,M,ID,IV,IER,AEPS)
!      IMPLICIT REAL*8(A-H,O,P,R-Z)
      IMPLICIT LOGICAL(Q)
      DIMENSION A(IA,N-M+1),ID(M+1),IV(N-M+1)
      PARAMETER(NIT=20)

      JF=1
      M1=M+1
      N1=N-M+1
      IER=0

!	The simplex iteration
      DO 4000 IT=1,NIT*(N+M)
!	Finding the minimum reduced cost coefficient
        RMIN=0.0
        K=0
        DO 2000 J=2,N1
          IF(A(JF,J).LT.RMIN) THEN
            RMIN=A(JF,J)
            K=J
          ELSE IF(IV(J).LE.M.AND.2-A(JF,J).LT.RMIN) THEN
            RMIN=2-A(JF,J)
            K=-J
          ENDIF
2000    CONTINUE
        IF(RMIN.GE.-AEPS) RETURN

!	Finding the pivot element
        K1=ABS(K)
        RMIN=0.0
        L=0
        DO 2400 J=2,M+1
          AJ=A(J,K1)
          IF(K.LT.0) AJ=-AJ
          IF(AJ.GT.AEPS) THEN
            R1=A(J,1)/AJ
            IF(R1.LT.RMIN.OR.L.EQ.0) THEN
              RMIN=R1
              L=J
            ENDIF
          ENDIF
2400    CONTINUE
        IF(L.EQ.0) THEN
!	The objective function is unbounded from below
          IER=63
          RETURN
        ENDIF

        IF(K.LT.0) THEN
          A(JF,K1)=-2+A(JF,K1)
          IV(K1)=-IV(K1)
        ENDIF

!	Exchange the variables
        L1=ID(L)
        ID(L)=IV(K1)
        IV(K1)=L1
        DO 3000 J=1,N1
          IF(J.NE.K1) THEN
            R1=A(L,J)/A(L,K1)
            DO 2800 I=1,M1
              IF(I.NE.L) A(I,J)=A(I,J)-A(I,K1)*R1
2800        CONTINUE
          ENDIF
3000    CONTINUE

        R1=ABS(A(L,K1))
        DO 3200 J=1,N1
          IF(J.NE.K1) A(L,J)=A(L,J)/R1
3200    CONTINUE
        DO 3400 I=1,M1
          IF(I.NE.L) A(I,K1)=-A(I,K1)/A(L,K1)
3400    CONTINUE
        A(L,K1)=1./R1
4000  CONTINUE

!	Iteration fails to converge
      IER=635
      END
